File: ios/Oriveo/Oriveo/Core/Models/ProviderLogoResolver.swift
import Foundation

enum ProviderLogoResolver {
    /// Inferring a relay's logo joins every enabled model's names and runs dozens of substring
    /// searches (about 10ms for 150 models), while the provider detail hero card reads it a dozen
    /// times per body and every providers write re-reads it. The cost sits in the system string
    /// search, so an optimized build is no faster. Remember the last inputs and result per provider
    /// and reuse them only when every input is equal; comparing `models` is O(1) while it still
    /// shares storage, and any content change recomputes.
    private struct Memo {
        let customName: String?
        let baseURLText: String?
        let relayKind: RelayKind?
        let models: [AIModel]
        let kind: ProviderKind

        func matches(_ provider: Provider) -> Bool {
            guard customName == provider.customName, baseURLText == provider.baseURLText else { return false }
            guard relayKind == provider.relayKind else { return false }
            return models == provider.models
        }
    }

    private static var memos: [UUID: Memo] = [:]
    private static let memoCapacity = 32

    #if DEBUG
    private static var computationCounts: [UUID: Int] = [:]

    /// DEBUG only: how many times a provider's logo was actually inferred (memo misses).
    static func computationCountForTesting(providerID: UUID) -> Int {
        computationCounts[providerID, default: 0]
    }
    #endif

    static func logoKind(for provider: Provider) -> ProviderKind {
        guard provider.kind == .relay else { return provider.kind }
        if let memo = memos[provider.id], memo.matches(provider) {
            return memo.kind
        }

        let kind = inferRelayLogoKind(for: provider)
        #if DEBUG
        computationCounts[provider.id, default: 0] += 1
        #endif
        if memos[provider.id] == nil, memos.count >= memoCapacity {
            memos.removeAll()
        }
        memos[provider.id] = Memo(
            customName: provider.customName,
            baseURLText: provider.baseURLText,
            relayKind: provider.relayKind,
            models: provider.models,
            kind: kind
        )
        return kind
    }

    /// Reads only displayName (for a relay it depends on customName and baseURLText), baseURLText,
    /// models and relayKind, which are exactly the fields `Memo.matches` compares. Change both
    /// together when adding an input.
    private static func inferRelayLogoKind(for provider: Provider) -> ProviderKind {
        let hints = ([provider.displayName, provider.baseURLText] + provider.models.flatMap { model in
            [model.groupKey, model.groupName, model.id, model.name]
        })
        .compactMap { $0 }
        .joined(separator: " ")
        .lowercased()

        if hints.isEmpty {
            return provider.relayKind.flatMap(logoKind(for:)) ?? .relay
        }

        if containsAny(hints, ["kimi", "moonshot", "moonshot.ai", "moonshot.cn"]) { return .moonshot }
        if containsAny(hints, ["grok", "xai", "x.ai"]) { return .grok }
        if containsAny(hints, ["mistral", "mixtral", "codestral", "magistral", "devstral", "ministral", "pixtral"]) { return .mistral }
        if containsAny(hints, ["openrouter"]) { return .openRouter }
        if containsAny(hints, ["openai", "gpt", "chatgpt"]) || containsAny(hints, [" o1", " o3", " o4"]) { return .openAI }
        if containsAny(hints, ["anthropic", "claude"]) { return .anthropic }
        if containsAny(hints, ["gemini", "google", "generativelanguage"]) { return .gemini }
        if containsAny(hints, ["deepseek"]) { return .deepseek }
        if containsAny(hints, ["qwen", "dashscope", "aliyun", "alibaba"]) { return .qwen }
        if containsAny(hints, ["groq"]) { return .groq }
        if containsAny(hints, ["together"]) { return .together }
        if containsAny(hints, ["fireworks"]) { return .fireworks }
        if containsAny(hints, ["minimax", "minimaxi"]) { return .miniMax }
        if containsAny(hints, ["zhipu", "z.ai", "bigmodel", "glm"]) { return .zhipu }
        if containsAny(hints, ["siliconflow"]) { return .siliconFlow }
        if containsAny(hints, ["oriveo"]) { return .openAI }

        return provider.relayKind.flatMap(logoKind(for:)) ?? .relay
    }

    private static func logoKind(for relayKind: RelayKind) -> ProviderKind? {
        switch relayKind {
        case .openaiCompatible, .codexStyle:
            return .openAI
        case .anthropicCompatible:
            return .anthropic
        case .geminiCompatible:
            return .gemini
        case .custom:
            return nil
        }
    }

    private static func containsAny(_ text: String, _ needles: [String]) -> Bool {
        needles.contains { text.contains($0) }
    }
}

File: android/app/src/main/java/ai/oriveo/community/core/model/ProviderLogoResolver.kt
package ai.oriveo.community.core.model

import androidx.annotation.VisibleForTesting
import java.lang.ref.WeakReference
import java.util.concurrent.atomic.AtomicInteger

/**
 * Infers the brand logo to show for a relay provider.
 *
 * An official kind answers itself; only a relay has to be guessed at, by concatenating the display
 * name, the base URL and the group key, group name, id and name of **every** model into one string
 * and running `contains` over it. A relay with many models produces tens of kilobytes, and the
 * `when` below performs close to forty substring scans, each one linear in that length.
 *
 * ## Why the in-process cache
 *
 * Provider cards and model pickers already wrap the call in `remember(provider)`, but cost
 * summaries run it on a background thread and the chat screen's model sheet calls it bare, so the
 * same provider is scanned several times within one pass.
 *
 * One entry is cached per provider id and a hit has to clear two gates:
 * 1. **The same `Provider` instance** (a weak reference compared by identity), which is the normal
 *    case when one render or one calculation asks about the same provider repeatedly. Not even the
 *    hint string has to be rebuilt. The reference is weak so the cache never pins a provider in
 *    memory.
 * 2. If the instance differs, the hints are rebuilt and compared **by content** — a re-emission
 *    from the database produces a new but identical instance, and this gate makes those pay for one
 *    hint build instead of forty substring scans.
 *
 * Only when both gates miss is a real scan performed. The branches of the `when` and their order
 * are unchanged; the cache only remembers the outcome of the same decision.
 */
fun resolveProviderLogoKind(provider: Provider): ProviderKind {
    if (provider.kind != ProviderKind.Relay) return provider.kind

    val cached = synchronized(relayLogoCache) { relayLogoCache[provider.id] }
    if (cached != null && cached.source.get() === provider) return cached.kind

    val fallback = provider.relayKind?.let(::relayKindLogoKind) ?: ProviderKind.Relay
    val hints = relayLogoHints(provider)
    if (hints.isEmpty()) return fallback
    if (cached != null && cached.hints == hints && cached.fallback == fallback) {
        synchronized(relayLogoCache) {
            relayLogoCache[provider.id] = cached.copyingSource(provider)
        }
        return cached.kind
    }

    relayLogoScanCount.incrementAndGet()
    val resolved = matchRelayLogoKind(hints) ?: fallback
    synchronized(relayLogoCache) {
        relayLogoCache[provider.id] = RelayLogoCacheEntry(WeakReference(provider), hints, fallback, resolved)
    }
    return resolved
}

private fun relayLogoHints(provider: Provider): String = buildList {
    add(provider.displayName)
    provider.baseUrlText?.let(::add)
    provider.models.forEach { model ->
        model.groupKey?.let(::add)
        model.groupName?.let(::add)
        add(model.id)
        add(model.name)
    }
}.joinToString(" ").lowercase()

/** Null means no brand matched, and the caller falls back to the relay kind. Branch order is load-bearing. */
private fun matchRelayLogoKind(hints: String): ProviderKind? {
    fun has(vararg needles: String): Boolean = needles.any { hints.contains(it) }

    return when {
        has("kimi", "moonshot", "moonshot.ai", "moonshot.cn") -> ProviderKind.Moonshot
        has("grok", "xai", "x.ai") -> ProviderKind.Grok

        has("mistral", "mixtral", "codestral", "magistral", "devstral", "ministral", "pixtral") -> ProviderKind.Mistral
        has("openrouter") -> ProviderKind.OpenRouter
        has("openai", "gpt", "chatgpt") || has(" o1", " o3", " o4") -> ProviderKind.OpenAI
        has("anthropic", "claude") -> ProviderKind.Anthropic
        has("gemini", "google", "generativelanguage") -> ProviderKind.Gemini
        has("deepseek") -> ProviderKind.DeepSeek
        has("qwen", "dashscope", "aliyun", "alibaba") -> ProviderKind.Qwen
        has("groq") -> ProviderKind.Groq
        has("together") -> ProviderKind.Together
        has("fireworks") -> ProviderKind.Fireworks
        has("minimax", "minimaxi") -> ProviderKind.MiniMax
        has("zhipu", "z.ai", "bigmodel", "glm") -> ProviderKind.Zhipu
        has("siliconflow") -> ProviderKind.SiliconFlow
        else -> null
    }
}

private class RelayLogoCacheEntry(
    val source: WeakReference<Provider>,
    val hints: String,
    val fallback: ProviderKind,
    val kind: ProviderKind,
) {
    fun copyingSource(provider: Provider) = RelayLogoCacheEntry(WeakReference(provider), hints, fallback, kind)
}

/** Sized after how many relays one person plausibly adds; beyond that the least recently used entry goes. */
private const val RELAY_LOGO_CACHE_MAX_ENTRIES = 64

private val relayLogoCache = object : LinkedHashMap<String, RelayLogoCacheEntry>(16, 0.75f, true) {
    override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, RelayLogoCacheEntry>): Boolean =
        size > RELAY_LOGO_CACHE_MAX_ENTRIES
}

/**
 * How many real scans happened; only a cache miss increments it. This counter is the one side
 * effect on the production path, and it exists because "the same fingerprint is scanned once" can
 * only be reported by the production code itself — a test that synthesised the number would be
 * proving nothing.
 */
@VisibleForTesting
internal val relayLogoScanCount = AtomicInteger(0)

@VisibleForTesting
internal fun resetRelayLogoCache() {
    synchronized(relayLogoCache) { relayLogoCache.clear() }
    relayLogoScanCount.set(0)
}

private fun relayKindLogoKind(relayKind: RelayKind): ProviderKind? = when (relayKind) {
    RelayKind.OpenAICompatible,
    RelayKind.CodexStyle,
    -> ProviderKind.OpenAI
    RelayKind.AnthropicCompatible -> ProviderKind.Anthropic
    RelayKind.GeminiCompatible -> ProviderKind.Gemini
    RelayKind.Custom -> null
}
